text/0000-canonical-metadata.md

```diff
@@ -1216,6 +1216,146 @@ interface StaticResourceReference {
+
+### TypeScript
+
+#### Root Metadata Object
+```ts
+interface BundleMetadata {
+    success: boolean;
+    version: string;
+    diagnostics: Array<Diagnostic>;
+    entry: string,
+    result: { [name: string]: FileMetadata } // metadata per each source
+    [api or interface]: ComponentInterface, // public methods, events, css properties, exports, slots.
+}
+
+interface BundleInterface {
+    type: 'module' | 'component',
+    attachedEvents: Array<Event>,
+    emittedEvents: Array<Event>,
+    publicProperties: Array<ClassMember>,
+    publicMethods: Array<ClassMember>,
+    slots: Array<Slot>,
+    exports: Array<ClassMetadata | FunctionMetadata | VariableMetadata>
+}
+```
+
+<details>
+<summary>Click to view format</summary>
+#### Root Metadata Object Per File
+```ts
+export interface FileMetadata {
+    path: string,
+    type: 'module' | 'component'| 'css',
+    dependencies: Array<Reference>, // dynamic import is treated as a module dependency type of 'dynamic'
+}
+```
+
+For every source in the bundle there will be a corresponding typed class which extends from the base FileMetadata.
+
+#### Script Metadata (.js|.ts)
+```ts
+interface ScriptMetadata extends FileMetadata {
+    declarations: Array<ClassMetadata | FunctionMetadata | VariableMetadata>
+} 
+
+interface ClassMetadata {
+   name: string,
+   extends: ExtendsMeta, // id, resource, location of the super 
+   classMembers: Array<ClassMember>, // includes private/public props/methods
+   documentation: ClassDocumentation,
+   events: Array<ClassEvent>,
+}
+
+interface ClassDocumentation {
+    classDescription: string,
+    html: string,
+    metadata: Object,
+}
+
+interface ExtendsMeta {
+    id: string,
+    type: Class | Function
+    resource: string,
+    location: Location,
+}
+
+interface ClassMember {
+    name: string,
+    type: string,
+    isPublic: boolean,
+    documentation?: ClassMemberDocumentation,
+    wire?: WireDependency
+}
+
+interface ClassMemberDocumentation {
+    description: string,
+    defaultValue: string,
+    dataType: string,
+    isRequired: boolean
+}
+
+interface ClassProperty {
+    hasGetter: boolean,
+    hasSetter: boolean,
+    value?: ClassMemberValue,
+}
+
+interface ClassMethod {
+    hasGetter: boolean,
+    hasSetter: boolean,
+    returnValue?: ClassMemberValue,
+}
+
+interface ClassMemberValue {
+    type: ClassMemberValueType, 
+    value: any,
+    importedName: string,
+}
+
+enum ClassMemberValueType {
+    ARRAY, BOOLEAN, MODULE, NUMBER, NULL, OBJECT, STRING, UNDEFINED, UNRESOLVED,
+}
+
+interface WireDependency {
+    adapter: WireTargetAdapter;
+    params?: { [name: string]: Value }; // value has type and actual value
+    
+}
+
+interface WireTargetAdapter {
+    name: string;
+    reference: string;
+}
+```
+
+### HTML Metadata
+```ts
+interface ComponentMetadata implements FileMetadata {
```

Comment:
Is this for the root HTML file or for an individual component referenced in the template?